### PR TITLE
Homogenize message structure

### PR DESCRIPTION
**Pull request by ghost**
Premise:
I know it's a lot of code but smaller pieces would have taken these changes out of context
*Unit test are all passed but now i can't test long chain of units interations 
*let me know if you find a problem or you disagree with my choices
*modifying classes i have tried to give credits to those who write the code i modified let me know if you think that i should add @ author YOURNAME
**@ author tag is in chronological order so my tag @ Mattia Fanan should be the last

Problems I've found and I've tried to solve:
--Messages structure were not homogeneous along the central chain of manager cause different needs in terms of number of peers

--SMSMessage promisquity (if I suddenly receive a SMSMessage out of context I can't know if I received it or I have to send it)
**Retieving phone number is a coin toss cause phone number isn't stored in sim but you get it from the net 

--Fragmented data flow along the central chain of manager caused by SmsManager.sendMessage taking two arguments 

--(minor problem I think it was resolved in previous code review)
in equals() instanceof(null) return false so o==null is pleonastic

How i have tried to solve them:
--I have extracted inhomogeneities from Message to Header so Messages can be treated easily having a unic interface for all possible (almost those i can think) variations
**also Header is very flexible having noPeers-onlySource-onlyDestinationbothPeers variations

--I have splitted SMSMessage in SMSMessageReceived and SMSMessageToSend
**SMSMessageReceived have an onlySourceHeader
**SMSMessageToSend have an onlyDestinationHeader
**doing this I extracted common code to AbstactSMSMessage

--I have implemented SMSHeader extracting common code to AbstractSMSHeader and extending it in:
**SMSReceivedHeader used by SMSMessageReceived 
**SMSToSendHeader used by SMSMessageToSend 

--Now that Messages have unique interface I have modyfied SMSMessageParse and ParseStrategy
**I have splitted ParseStrategy in DataParseStrategy and MessageParseStrategy obtaining more flexibility

--I have standardized data flow
**Now ParserStartegy mapp between single data entity (string in this case) and Message
**Fragmentatin introduced from SmsManager.sendMessage is now confined in APIMessage and APIParser far from the central chain of Message managers
(can be removed in future building and retrieving raw sms PDU)

--I have also tried to unify javadoc style takink as reference Luca Crema's javadoc and applying changes according to:
https://www.oracle.com/technetwork/java/javase/documentation/index-137868.html

### CODE REVIEW

**Comment by CremaLuca** [comment]
Sorry man, I'm not reviewing 43 files all at once

**Comment by CremaLuca** [comment]
> Messages structure were not homogeneous along the central chain of manager cause different needs in terms of number of peers
What? Where? What do you mean I genuinely can't understand sorry

**Comment by CremaLuca** [comment]
> SMSMessage promisquity

It has never been a problem, it's pretty clear that if the message is a param on received listener it's a received message. Did you find a situation where this was not **that** clear and you needed to have both the sender and the receiver (given that getting your own phone number is a Russian roulette)?

**Comment by CremaLuca** [comment]
> Fragmented data flow along the central chain of manager caused by SmsManager.sendMessage taking two arguments

Again, what do you mean with fragmented? And `SmsManager.sendMessage` takes 5 arguments

**Comment by CremaLuca** [comment]
> in equals() instanceof(null) return false so o==null is pleonastic

Noice, correct, well spotted

**Comment by CremaLuca** [comment]
> I have splitted SMSMessage in SMSMessageReceived and SMSMessageToSend

This looks a lot like over-engineering, you're making multiple models of reality that could be merged into the same (maybe less detailed) object

File: smslibrary/src/main/java/com/eis/communication/CommunicationHandler.java [hunk 1]
```diff
@@ -9,7 +9,8 @@
 
     /**
      * Sends a single message in the channel, message content must be shorter than the maximum channel message size
-     * @param message message to be sent in the channel to a peer
+     *
+     * @param message the message to be sent in the channel to a peer
```
Comment (CremaLuca) [inline smslibrary/src/main/java/com/eis/communication/CommunicationHandler.java]: You see, this is a correct edit to the code, I would have accepted it right away, but it's been edited with other 42 files and it's going to take a long time before it gets into the master code (if it's getting merged). You could have pulled changes like these in another request, and leave to this request the essential

File: smslibrary/src/main/java/com/eis/communication/Peer.java [hunk 2]
```diff
@@ -1,17 +1,18 @@
...
  * Represents a user in the network
+ *
  * @param <A> The type of address used by the network
+ * @see Serializable
  */
 public interface Peer<A> extends Serializable {
 
     /**
      * Retrieves the peer's address
+     *
```
Comment (CremaLuca) [inline smslibrary/src/main/java/com/eis/communication/Peer.java]: I mean, look at this, another edited file just to add some documentation, could have been put in another pull request

**Comment by MattiaFanan-uni** [comment]
[Message inhomogeneity .pdf](https://github.com/Cogno-IDU/smslibrary/files/3940140/Message.inhomogeneity.pdf)

**Comment by MattiaFanan-uni** [comment]
[Data flow fragmentation.pdf](https://github.com/Cogno-IDU/smslibrary/files/3940146/Data.flow.fragmentation.pdf)

File: smslibrary/src/main/java/com/eis/communication/Peer.java (hunk 2 above)
Comment (MattiaFanan-uni) [reply smslibrary/src/main/java/com/eis/communication/Peer.java]: I know but “già che c’ero”

**Comment by CremaLuca** [comment]
About in-homogenity: I couldn't really think of a situation where I need both sender and receiver peer for a message, the type of messages are already well identified by their position in the code and I think you're trying to over-modeling a reality that right now is already defined **good enough** (I'm not saying it's the right or best solution, I think right now it's simple and requires only a small time to understand how it works).
Also what's the meaning of `Message<Header>`? I can't have a clear idea of what that'd mean, I could really use an explanation because right now I'm thinking that a message (like SMS) could have multiple type of headers and... Idk doesn't seem what you actually mean.

**Comment by MattiaFanan-uni** [comment]
In the net layer is not 100% sure the a message I have received is directed to me
(Except for big simplifications that we can do)

**Comment by MattiaFanan-uni** [comment]
Message<Header> means that you could write code for all generics messages or a subset but you have no immediate limitation(also now you don’t have limitation) nor future possible limitation

**Review by albertoursino — COMMENTED** [review]
What do we have to do with this pull request?
